### PR TITLE
add example code

### PR DESCRIPTION
**Pull request by mhenPiyawut**

### CODE REVIEW

File: mock_ros2_project/publisher_node.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+class PublisherNode(Node):
+    """Simple publisher node that publishes string messages."""
+    
+    def __init__(self):
+        super().__init__('publisher_node')
+        
+        # Create publisher
+        self.publisher_ = self.create_publisher(String, 'topic', 10)
+
```
Comment (mhenPiyawut) [inline mock_ros2_project/publisher_node.py:22]: Expose topic name and QoS depth using declare_parameter so users can configure them at launch.

File: mock_ros2_project/action_client.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    def __init__(self):
+        super().__init__('action_client')
+        
+        # Create action client
+        self._action_client = ActionClient(self, Fibonacci, 'fibonacci')
+
+    def send_goal(self, order):
+        """Send a goal to the action server."""
+        goal_msg = Fibonacci.Goal()
+        goal_msg.order = order
```
Comment (mhenPiyawut) [inline mock_ros2_project/action_client.py:28]: wait_for_server() without a timeout can block forever if the server is unavailable; supply a timeout or use the async variant.

File: mock_ros2_project/action_server.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        feedback_msg = Fibonacci.Feedback()
+        feedback_msg.partial_sequence = [0, 1]
+        
+        for i in range(1, goal_handle.request.order):
+            feedback_msg.partial_sequence.append(
+                feedback_msg.partial_sequence[i] + feedback_msg.partial_sequence[i-1]
+            )
+            
+            self.get_logger().info(f'Feedback: {feedback_msg.partial_sequence}')
+            goal_handle.publish_feedback(feedback_msg)
```
Comment (mhenPiyawut) [inline mock_ros2_project/action_server.py:45]: Handle cancellation inside the loop (goal_handle.is_cancel_requested()) and terminate early to comply with ROS2 action semantics.

File: mock_ros2_project/publisher_node.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    def __init__(self):
+        super().__init__('publisher_node')
+        
+        # Create publisher
+        self.publisher_ = self.create_publisher(String, 'topic', 10)
+        
+        # Create timer to publish messages periodically
+        timer_period = 0.5  # seconds
+        self.timer = self.create_timer(timer_period, self.timer_callback)
+
```
Comment (mhenPiyawut) [inline mock_ros2_project/publisher_node.py:26]: Timer period should be a declared parameter to avoid hard-coding publish frequency.

File: mock_ros2_project/action_server.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        
+        self.get_logger().info('Action server ready to generate Fibonacci sequences.')
+
+    def execute_callback(self, goal_handle):
+        """Execute callback for the Fibonacci action."""
+        self.get_logger().info('Executing goal...')
+        
+        feedback_msg = Fibonacci.Feedback()
+        feedback_msg.partial_sequence = [0, 1]
+
```
Comment (mhenPiyawut) [inline mock_ros2_project/action_server.py:38]: Initialising with [0,1] ignores requests with order < 2; add input validation or tailor the sequence size.

File: mock_ros2_project/service_server.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+def main(args=None):
+    """Main function to run the service server."""
+    rclpy.init(args=args)
+    
+    service_server = ServiceServer()
+    
+    try:
+        rclpy.spin(service_server)
+    except KeyboardInterrupt:
+        pass
```
Comment (mhenPiyawut) [inline mock_ros2_project/service_server.py:50]: Silently swallowing KeyboardInterrupt hides shutdown reason; log a friendly shutdown message instead.

File: mock_ros2_project/action_client.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        
+        self._send_goal_future = self._action_client.send_goal_async(
+            goal_msg,
+            feedback_callback=self.feedback_callback
+        )
+        
+        self._send_goal_future.add_done_callback(self.goal_response_callback)
+
+    def goal_response_callback(self, future):
+        """Handle goal response from action server."""
```
Comment (mhenPiyawut) [inline mock_ros2_project/action_client.py:40]: Check future.exception() before accessing result to handle send_goal errors cleanly.

File: mock_ros2_project/publisher_node.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        self.get_logger().info('Publisher node has been started.')
+
+    def timer_callback(self):
+        """Timer callback function that publishes messages."""
+        msg = String()
+        msg.data = f'Hello World: {self.i}'
+        
+        self.publisher_.publish(msg)
+        self.get_logger().info(f'Publishing: "{msg.data}"')
+
```
Comment (mhenPiyawut) [inline mock_ros2_project/publisher_node.py:39]: Use throttled logging (`self.get_logger().info_throttle`) or `debug` level to prevent log spam at 2 Hz.

File: mock_ros2_project/action_server.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        for i in range(1, goal_handle.request.order):
+            feedback_msg.partial_sequence.append(
+                feedback_msg.partial_sequence[i] + feedback_msg.partial_sequence[i-1]
+            )
+            
+            self.get_logger().info(f'Feedback: {feedback_msg.partial_sequence}')
+            goal_handle.publish_feedback(feedback_msg)
+            
+            time.sleep(1)  # Simulate work
+
```
Comment (mhenPiyawut) [inline mock_ros2_project/action_server.py:48]: time.sleep() blocks the single-threaded executor; use rclpy.rate or a WallTimer for cooperative spinning.

File: mock_ros2_project/action_client.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            return
+        
+        self.get_logger().info('Goal accepted :)')
+        
+        self._get_result_future = goal_handle.get_result_async()
+        self._get_result_future.add_done_callback(self.get_result_callback)
+
+    def get_result_callback(self, future):
+        """Handle result from action server."""
+        result = future.result().result
```
Comment (mhenPiyawut) [inline mock_ros2_project/action_client.py:54]: Calling rclpy.shutdown() from inside a callback can halt the executor unexpectedly; trigger shutdown from main or after spin terminates instead.

File: mock_ros2_project/service_client.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    def __init__(self):
+        super().__init__('service_client')
+        
+        # Create client
+        self.cli = self.create_client(AddTwoInts, 'add_two_ints')
+        
+        # Wait for service to be available
+        while not self.cli.wait_for_service(timeout_sec=1.0):
+            self.get_logger().info('Service not available, waiting again...')
+
```
Comment (mhenPiyawut) [inline mock_ros2_project/service_client.py:27]: Add `and rclpy.ok()` to the loop to break gracefully on shutdown instead of potentially hanging.

File: mock_ros2_project/subscriber_node.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    def __init__(self):
+        super().__init__('subscriber_node')
+        
+        # Create subscription
+        self.subscription = self.create_subscription(
+            String,
+            'topic',
+            self.listener_callback,
+            10
+        )
```
Comment (mhenPiyawut) [inline mock_ros2_project/subscriber_node.py:26]: Prefer passing an explicit `QoSProfile` (e.g., `QoSProfile(depth=10)`) to avoid implicit defaults and to configure reliability/durability upfront.

File: mock_ros2_project/service_client.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        while not self.cli.wait_for_service(timeout_sec=1.0):
+            self.get_logger().info('Service not available, waiting again...')
+        
+        self.req = AddTwoInts.Request()
+
+    def send_request(self, a, b):
+        """Send a request to the service."""
+        self.req.a = a
+        self.req.b = b
+
```
Comment (mhenPiyawut) [inline mock_ros2_project/service_client.py:34]: Using one persistent `Request` instance risks shared-state issues; instantiate a new `AddTwoInts.Request()` inside `send_request()`.

File: mock_ros2_project/subscriber_node.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        self.subscription  # prevent unused variable warning
+
```
Comment (mhenPiyawut) [inline mock_ros2_project/subscriber_node.py:28]: This extra attribute access is not required; `self.subscription` is already stored and will not trigger F841 warnings.

File: mock_ros2_project/service_client.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    def send_request(self, a, b):
+        """Send a request to the service."""
+        self.req.a = a
+        self.req.b = b
+        
+        self.future = self.cli.call_async(self.req)
+        rclpy.spin_until_future_complete(self, self.future)
+        
+        return self.future.result()
```
Comment (mhenPiyawut) [inline mock_ros2_project/service_client.py:38]: `spin_until_future_complete()` lacks a timeout, so the node could block forever if the server never responds; consider a timeout or an async callback.

**Review by mhenPiyawut — COMMENTED** [review]
Critical Issues → rclpy.shutdown is invoked inside a callback, which can prematurely tear down the entire context while other callbacks/executors are active. Improvements → wait_for_server() is used synchronously without a timeout risking indefinite block, and goal_response_callback does not guard against future exceptions. Nitpicks → none. Tests & Docs → add tests for missing server and rejected goals, document shutdown flow.

**Review by mhenPiyawut — COMMENTED** [review]
Critical Issues → execute_callback never checks goal_handle.is_cancel_requested(), so the server cannot honor client cancellation requests.
Improvements → Always validate goal_handle.request.order and adapt the initial sequence to correctly satisfy orders < 2.
Nitpicks → Using time.sleep blocks the executor; prefer rclpy.sleep or a Timer for non-blocking feedback loops.
Tests & Docs → Add tests for cancellation and edge-case orders (0 and 1).

**Review by mhenPiyawut — COMMENTED** [review]
Improvements: Make topic name, QoS depth, and timer period configurable via declare_parameter to increase reusability → Consider throttling or changing logger level in fast timer callback to avoid flooding.
Nitpicks: Variable `i` could be renamed to something descriptive.
Tests & Docs: Add simple launch test verifying message count over time.

**Review by mhenPiyawut — COMMENTED** [review]
No critical issues → Consider logging shutdown event instead of silent pass → Nitpick: QoS settings not specified but default is usually fine → Tests/Docs: Example sufficiently documented.

**Review by mhenPiyawut — COMMENTED** [review]
Critical Issues → None.
Improvements → 1) Reuse of a single request instance can cause accidental state sharing; create a fresh request each call. 2) spin_until_future_complete without timeout may block indefinitely; consider timeout or asynchronous callback. 3) Loop waiting for service does not exit on shutdown; check rclpy.ok() to avoid hanging during Ctrl+C.
Nitpicks → Handle argv parsing errors.
Tests & Docs → Add tests covering timeout and multiple consecutive calls.

**Review by mhenPiyawut — COMMENTED** [review]
Improvements: 1) Use an explicit QoSProfile when creating the subscription for clearer reliability/durability control instead of the legacy depth-only form. 2) Remove the redundant `self.subscription` line; properties are not flagged by F841 so this extra access is unnecessary. Nitpicks: Add type hints for `msg` parameter. Tests & Docs: Example is self-explanatory, no further docs needed.
